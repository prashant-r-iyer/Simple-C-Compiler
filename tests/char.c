
char* g;

void main()
{
	char* h;
	h = "abcd";

	printf("h[1]=%c\n", h[1]);

	g = "efg";

	printf("g[3]=%c\n", g[2]);
}

