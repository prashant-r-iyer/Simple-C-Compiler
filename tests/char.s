	.data
	.comm g, 8
	.text

.globl main
main:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9
	movq $string0, %rbx
	movq %rbx, -8(%rbp)
	movq $string1, %rbx

	# push 1
	movq $1, %r10
	addq -8(%rbp), %r10
	movq $0, %r13
	movb (%r10), %r13b
	movq %r13, %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	movq $string2, %rbx
	movq %rbx, g
	movq $string3, %rbx

	# push 2
	movq $2, %r10
	addq g, %r10
	movq $0, %r13
	movb (%r10), %r13b
	movq %r13, %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

string0:
	.string "abcd"

string1:
	.string "h[1]=%c\n"

string2:
	.string "efg"

string3:
	.string "g[3]=%c\n"

