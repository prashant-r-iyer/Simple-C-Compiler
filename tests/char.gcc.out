char.c: In function ‘main’:
char.c:9:9: warning: implicit declaration of function ‘printf’ [-Wimplicit-function-declaration]
    9 |         printf("h[1]=%c\n", h[1]);
      |         ^~~~~~
char.c:1:1: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
  +++ |+#include <stdio.h>
    1 | 
char.c:9:9: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
    9 |         printf("h[1]=%c\n", h[1]);
      |         ^~~~~~
char.c:9:9: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
